Handle aneuploid likelihoods

ronlevine opened the pull request:
Implements https://github.com/samtools/htsjdk/issues/437.

Previously, only diploid allele pair indices could be accessed by PL index:
`GenotypeLikelihoodsAllelePair getAllelePair(final int PLindex)`

This pull request adds the ability to get the allele indices for any ploidy:
`List<Integer> GenotypeLikelihoods.getAlleles(final int PLindex, final int ploidy)`

ronlevine commented:
@yfarjoun Please review.

ronlevine commented:
@yfarjoun Please review if you can.

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -101,11 +108,11 @@ private GenotypeLikelihoods(double[] asVector) {
     /**
-     * Returns the genotypes likelihoods in negative log10 vector format.  pr{AA} = x, this
+     * The genotypes likelihoods in negative log10 vector format.  pr{AA} = x, this
      * vector returns math.log10(x) for each of the genotypes.  Can return null if the
```
seems like it returns -10*log10(x), right?

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -318,6 +325,12 @@ public GenotypeLikelihoodsAllelePair(final int alleleIndex1, final int alleleInd
+     * @param altAlleles   number of alternate alleles
+     * @return cache of diploid alleles for each PL index
+     */
     private static GenotypeLikelihoodsAllelePair[] calculatePLcache(final int altAlleles) {
```
might as well rename this function (its private anyhow) to calculateDiploidPLcache

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -338,6 +351,29 @@ public GenotypeLikelihoodsAllelePair(final int alleleIndex1, final int alleleInd
+     * @param ploidy        Number of chromosomes in set
+     * @param genotype      An entry of ploidy alleles
+     */
+    protected static void calculatePLcache(final int numAlleles, final int ploidy, final List<Integer> genotype){
```
and rename this one calculateAnyPloidyPLcache (BTW, the thing you are refering to is spelled 'aneuploid' or 'aneuploidy' I'm specifically using the spelling above to indicate 'any ploidy'...i.e. not only diploid.

yfarjoun replied:
protect against someone putting genotype with .size()>ploidy, as that will not terminate.

yfarjoun replied:
space

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -338,6 +351,29 @@ public GenotypeLikelihoodsAllelePair(final int alleleIndex1, final int alleleInd
+    protected static void calculatePLcache(final int numAlleles, final int ploidy, final List<Integer> genotype){
+        if (genotype.size() == ploidy) {
+            plIndexToAlleleIndices.add(genotype);
+        }
```
please use } else { on the same line.

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -338,6 +351,29 @@ public GenotypeLikelihoodsAllelePair(final int alleleIndex1, final int alleleInd
+        }
+        else {
+            for (int a=0; a < numAlleles; a++) {
+                List gt = new ArrayList<Integer>(Arrays.asList(a));
```
why is gt not a List<Integer>?

ronlevine replied:
Isn't it a List?

yfarjoun replied:
sorry, markdown ate my comment:
why is gt not a `List<Integer>`?

ronlevine replied:
Because my backspace key ate my `<Integer>`. ;-)

yfarjoun replied:
Use Collections.SingletonList(a) instead of asList?

ronlevine replied:
Yes. Safer to make it immutable and preface it with a "final".

ronlevine replied:
Actually, gt is mutable since the next statement adds `genotype` so will make it `final List<Integer>`.

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -338,6 +351,29 @@ public GenotypeLikelihoodsAllelePair(final int alleleIndex1, final int alleleInd
+    /**
+     * Calculate the cache of anyploid alleles for each PL index
+     * Creates the ordering for all possible combinations of ploidy alleles. Computed recursively and the
+     * result is stored in a cache.
```
could you include the recursive formula you are using and a reference for where it comes from?

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -338,6 +351,29 @@ public GenotypeLikelihoodsAllelePair(final int alleleIndex1, final int alleleInd
+        else {
+            for (int a=0; a < numAlleles; a++) {
+                List gt = new ArrayList<Integer>(Arrays.asList(a));
+                gt.addAll(genotype);
```
looks like `gt` is a simple copy of `genotype`, was that intended?

yfarjoun replied:
I'm slowly getting the combinatorial construction you are doing here....bear with me...

ronlevine replied:
This will help: refer to "Algorithm for enumerating the genotypes given ploidy and alleles" in  http://genome.sph.umich.edu/wiki/Relationship_between_Ploidy,_Alleles_and_Genotypes

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -72,6 +72,13 @@
     */
     private final static GenotypeLikelihoodsAllelePair[] PLIndexToAlleleIndex = calculatePLcache(MAX_ALT_ALLELES_THAT_CAN_BE_GENOTYPED);
```
change the name of the static int to something that references 'diploid'

yfarjoun replied:
nm. I guess it's the same limit on non-diploid.

yfarjoun replied:
also, it would be good to add "diploid" to the variable itself.

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -338,6 +351,29 @@ public GenotypeLikelihoodsAllelePair(final int alleleIndex1, final int alleleInd
+     */
+    protected static void calculatePLcache(final int numAlleles, final int ploidy, final List<Integer> genotype){
+        if (genotype.size() == ploidy) {
+            plIndexToAlleleIndices.add(genotype);
```
perhaps add a comment here 
`// and we are done with the calculation`
so help the poor reader...

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -72,6 +72,13 @@
+    /*
+    * a cache of the PL index to the anyploid alleles it represents over all possible numbers of alternate alleles
+    */
+    protected final static List<List<Integer>> plIndexToAlleleIndices = new ArrayList<List<Integer>>();
```
This should be a `Map<Integer, List<Integer>>`, not a `List<List<Integer>>`, no? That way it is clear that you put something in the right place. regardless of order of insertion. Also it is much clearer then that the map is ploidy->plIndexToAlleleIndeciesList(of that ploidy)

yfarjoun replied:
scratch that. I got it wrong. it's not at all what I though...and I don't like it! The problem is that one cannot deal with two different ploidies at the same time since this is a singleton object I think that this should be encapsulated in a map (`Map<Integer,List<List<Integer>>>`) or an inner class so that one could have multiple instances of NonDiplodGenotypeLikelihoods with different ploidies...somewhat like LogMath in MathUtil.

ronlevine replied:
How about keeping the data structure as is but lazily initializing it with the double checked lock pattern in `initializeAneuploidPLIndexToAlleleIndices(final int altAlleles, final int ploidy))`? `aneuploidPLIndexToAlleleIndices` will only exist for one ploidy and the operations will be thread safe.

yfarjoun replied:
So you'll also store a static `ploidy` variable that will tell you for what ploidy the last is initialized and if it doesn't match, you'll reinitialize it? is that the idea? I'm not sure what the advantage is. Using a map will be thread safe will be just as "lazy" and allow for the odd case where two different ploidies are requested to remain as efficient. happy to be convinced otherwise...just interested in hearing the reasoning.

ronlevine replied:
Yes, the ploidy would have to be saved. My logic is to only deal with one ploidy. With the `map`, the application developer could initialize for multiple ploidys, which would waste memory.

yfarjoun replied:
well, if the developer needs more than one ploidy, what else should they
do? if they do it without needing, that's their fault. I would put a
comment in the initialization function that makes it clear that one should
only initialize ploidies that one is interested in...



On Sat, Feb 20, 2016 at 6:12 PM, Ron Levine <notifications@github.com>
wrote:

> In src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java
> <https://github.com/samtools/htsjdk/pull/466#discussion_r53559062>:
>
> > @@ -72,6 +72,13 @@
> >      */
> >      private final static GenotypeLikelihoodsAllelePair[] PLIndexToAlleleIndex = calculatePLcache(MAX_ALT_ALLELES_THAT_CAN_BE_GENOTYPED);
> >
> > +
> > +    /*
> > +    * a cache of the PL index to the anyploid alleles it represents over all possible numbers of alternate alleles
> > +    */
> > +    protected final static List<List<Integer>> plIndexToAlleleIndices = new ArrayList<List<Integer>>();
>
> Yes, the ploidy would have to be saved. My logic is to only deal with one
> ploidy. With the map, the application developer could initialize for
> multiple ploidys, which would waste memory.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/466/files#r53559062>.
>

ronlevine replied:
:+1: I'll go with that.

yfarjoun commented on src/tests/java/htsjdk/variant/variantcontext/GenotypeLikelihoodsUnitTest.java:
```diff
@@ -182,22 +187,122 @@ public void testCalculatePLindex(){
+
+    @Test(dataProvider = "testCalculatePLcacheData")
+    public void testCalculatePLcache(final int numAlleles, final int ploidy,  final List<List<Integer>> expected){
+        GenotypeLikelihoods.plIndexToAlleleIndices.clear();
```
Test should be using `initializePLIndexToAlleleIndices`, not replecating the code.

yfarjoun commented on src/tests/java/htsjdk/variant/variantcontext/GenotypeLikelihoodsUnitTest.java:
```diff
@@ -182,22 +187,122 @@ public void testCalculatePLindex(){
+    @DataProvider
+    public Object[][] testGetAllelesData() {
+        return new Object[][]{
+                {0, 2, 3, Arrays.asList(0,0,0)},
```
spaces after commas in lists

yfarjoun commented on src/tests/java/htsjdk/variant/variantcontext/GenotypeLikelihoodsUnitTest.java:
```diff
@@ -182,22 +187,122 @@ public void testCalculatePLindex(){
-    private void allelePairTest(int PLindex, int allele1, int allele2) {
+
+    @Test(dataProvider = "testGetAllelePairData")
+    public void testGetAllelePair(final int PLindex, final int allele1, final int allele2){
```
space between `)` and `{`

yfarjoun commented on src/tests/java/htsjdk/variant/variantcontext/GenotypeLikelihoodsUnitTest.java:
```diff
@@ -182,22 +187,122 @@ public void testCalculatePLindex(){
+    }
+
+    @Test(dataProvider = "testCalculatePLcacheData")
+    public void testCalculatePLcache(final int numAlleles, final int ploidy,  final List<List<Integer>> expected){
```
space here too.

yfarjoun commented on src/tests/java/htsjdk/variant/variantcontext/GenotypeLikelihoodsUnitTest.java:
```diff
@@ -182,22 +187,122 @@ public void testCalculatePLindex(){
+    public void testCalculatePLcache(final int numAlleles, final int ploidy,  final List<List<Integer>> expected){
+        GenotypeLikelihoods.plIndexToAlleleIndices.clear();
+        GenotypeLikelihoods.calculatePLcache(numAlleles, ploidy, new ArrayList<Integer>());
+        for ( int i=0; i < GenotypeLikelihoods.plIndexToAlleleIndices.size() ; i++){
```
space

yfarjoun commented on src/tests/java/htsjdk/variant/variantcontext/GenotypeLikelihoodsUnitTest.java:
```diff
@@ -182,22 +187,122 @@ public void testCalculatePLindex(){
+    }
+
+    @Test(dataProvider = "testGetAllelesData")
+    public void testGetAlleles(final int PLindex, final int numAltAlleles, final int ploidy, final List<Integer> expected ){
```
space

yfarjoun commented on src/tests/java/htsjdk/variant/variantcontext/GenotypeLikelihoodsUnitTest.java:
```diff
@@ -182,22 +187,122 @@ public void testCalculatePLindex(){
+    }
+
+    @Test(dataProvider = "testGetAllelesIndexOutOfBoundsData", expectedExceptions = IndexOutOfBoundsException.class)
+    public void testGetAllelesOutOfBounds(final int PLindex, final int numAltAlleles, final int ploidy){
```
space

yfarjoun commented on src/tests/java/htsjdk/variant/variantcontext/GenotypeLikelihoodsUnitTest.java:
```diff
@@ -182,22 +187,122 @@ public void testCalculatePLindex(){
+    }
+
+    @Test(expectedExceptions = IllegalStateException.class)
+    public void testGetAllelesUnitialized(){
```
space

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -422,6 +459,43 @@ public static GenotypeLikelihoodsAllelePair getAllelePair(final int PLindex) {
+     * @param altAlleles number of alternate alleles
+     * @param ploidy    number of chromosomes
+     */
+    public static synchronized void initializePLIndexToAlleleIndices(final int altAlleles, final int ploidy){
```
space

yfarjoun commented on src/tests/java/htsjdk/variant/variantcontext/GenotypeLikelihoodsUnitTest.java:
```diff
@@ -29,12 +29,17 @@
 // the imports for unit testing.
 
 
+import com.sun.glass.ui.SystemClipboard;
```
SystemClipboard?

yfarjoun commented on src/tests/java/htsjdk/variant/variantcontext/GenotypeLikelihoodsUnitTest.java:
```diff
@@ -29,12 +29,17 @@
 
 
+import com.sun.glass.ui.SystemClipboard;
+import com.sun.tools.javac.api.ClientCodeWrapper;
```
I think you have a few extra import statements

ronlevine replied:
Y. InelliJ black magiic

yfarjoun commented:
back to you @ronlevine. please do not rebase your next commit.

ronlevine commented:
Isn't aneuploidy "the presence of an abnormal number of chromosomes in a cell"? I was using anyploidy to be equivalent to general ploidy.

yfarjoun commented:
yes. in that case: CamelCase please.

On Wed, Feb 17, 2016 at 10:19 PM, Ron Levine <notifications@github.com>
wrote:

> Isn't aneuploidy "the presence of an abnormal number of chromosomes in a
> cell"? I was using anyploidy to be equivalent to general ploidy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/466#issuecomment-185525104>.
>

ronlevine commented:
@yfarjoun Please review. I think I addressed all of your comments. Note that I changed the allele ordering (calculateAneuploidPLcache()) to be consistent with the VCF 4.3 spec. Please comment if the accompanying doc is not clear or sufficient.

ronlevine commented:
@yfarjoun Comments addressed. Please review.

ronlevine commented:
@yfarjoun Please review. Responded to comments and made code/comments more legible.

ronlevine commented:
@yfarjoun Please take another look.

yfarjoun commented on src/java/htsjdk/variant/variantcontext/GenotypeLikelihoods.java:
```diff
@@ -338,6 +353,59 @@ public GenotypeLikelihoodsAllelePair(final int alleleIndex1, final int alleleInd
+            final List<Integer> gt = new ArrayList<Integer>(Arrays.asList(a));
+            gt.addAll(genotype);
+            if ( ploidy == 1 ) // have all ploidy alleles for a PL index
+                anyploidPLIndexToAlleleIndices.add(gt);
```
sorry, missed this before: multiple lines, please use braces:
```
if () {
...
} else {
..
}

yfarjoun commented:
thanks @ronlevine for this. minor comment and then :+1:  for rebase/squash/merge.
